cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/LogParserTransform.java

```diff
@@ -0,0 +1,183 @@
+
+import java.nio.ByteBuffer;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Parses log formats to extract access information
+ */
+@Plugin(type = "transform")
+@Name("LogParser")
+@Description("Parses logs for relevant information.")
+public class LogParserTransform extends Transform<StructuredRecord, StructuredRecord> {
+  private static final Schema LOG_SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("uri", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("ip", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("browser", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("device", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG))
+  );
+  private static final String LOG_FORMAT_DESCRIPTION = "Log format to parse. Currently only supports S3 and " +
+    "CLF Log format.";
+  private static final String INPUT_TYPE_DESCRIPTION = "Type of the value in the input schema which encodes the " +
+    "log information. Must be either \"string\" or \"bytes\"";
+  private static final String INPUT_NAME_DESCRIPTION = "Name of the value in the input schema which encodes the " +
+    "log information.";
+  private static final Logger LOG = LoggerFactory.getLogger(LogParserTransform.class);
+  private static final Pattern CLF_LOG_PATTERN = Pattern.compile(
+    //   IP                    id    user      date          request     code     size    referrer    user agent
+    "^([\\d.]+|[:][:][\\d]) (\\S+) (\\S+) \\[([^\\]]+)\\] \"([^\"]+)\" (\\d{3}) ([-\"\\d]+) \"([^\"]+)\" \"([^\"]+)\"");
+  private static final Pattern S3_LOG_PATTERN = Pattern.compile(
+    // bucket owner name   time           ip                    req   reqID operation  key     URI
+    "^(\\S+) (\\S+) \\[(\\p{Print}+)\\] ([\\d.]+|[:][:][\\d]) (\\S+) (\\S+) (\\S+) (\\S+) \"([^\"]+)\" " +
+      //  HTTP stat error code  bytes sent     obj size  time    turn time  referrer         user agent
+      "(\\d{3}) (\\p{Print}+) ([-\"\\d]+) ([-\"\\d]+) ([\\d]+) ([-\"\\d]+) \"(\\p{Print}+)\" \"([^\"]+)\" " +
+      //  version id
+      "(\\p{Print}+)");
+  private static final Pattern REQUEST_PAGE_PATTERN = Pattern.compile("(\\S+)\\s(\\S+).*");
+  private static final int[] S3_INDICES = {9, 3, 4, 17};
+  private static final int[] CLF_INDICES = {5, 4, 1, 9};
+  private static final int S3_REGEX_LENGTH = 18;
+  private static final int CLF_REGEX_LENGTH = 9;
+  private final LogParserConfig config;
+
+  public LogParserTransform(LogParserConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    String log;
+    if ("string".equals(config.inputType)) {
+      log = input.get(config.inputName);
+    } else if ("bytes".equals(config.inputType)) {
+      log = Bytes.toString((ByteBuffer) input.get(config.inputName));
```

Comment:
Hmmm, so this transform can only be applied to a stream source? Or something that sends a ``ByteBuffer``. If so, that should be documented in the description.

However, I think it would be better if you could

```
Object data = input.get(config.inputName);
if (data instanceof ByteBuffer) {
  // cast to ByteBuffer
} else if (data instanceof byte[]) {
  // cast to byte[]
}
```